snooty/diagnostics.py

```diff
@@ -0,0 +1,204 @@
+        else:
+            start_line, start_column = start
+        self.start = (start_line, start_column)
+
+        if end is None:
+            end_line, end_column = start_line, 1000
+        elif isinstance(end, int):
+            end_line, end_column = end, 1000
+        else:
+            end_line, end_column = end
+        self.end = (end_line, end_column)
+
+    class Level(enum.IntEnum):
+        info = 0
+        error = 1
+        warning = 2
+
+        @classmethod
+        def from_docutils(cls, docutils_level: int) -> "Diagnostic.Level":
+            level = docutils_level - 1
+            level = min(level, cls.warning)
+            level = max(level, cls.info)
+            return cls(level)
+
+    @property
+    def severity(self) -> "Diagnostic.Level":
+        raise TypeError("Cannot access the severity of an abstract base Diagnostic")
+
+    @property
+    def severity_string(self) -> str:
+        return self.severity.name.title()
+
+
+class UnexpectedIndentation(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class InvalidURL(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+# this could captuer 'literal include expected path arg..as well?
+class ExpectedPathArg(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        name: Path,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(f'"{name}" expected a path argument', start, end)
+        self.name = name
+
+
+class ExpectedImgArg(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class ImgExpectedButNotRequired(Diagnostic):
+    severity = Diagnostic.Level.warning
+
+
+class OptionsNotSupported(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class GitMergeConflictArtifactFound(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        path: Path,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(
+            f"Git merge conflict artifact found in {str(path)} on line {str(start)}",
+            start,
+            end,
+        )
+        self.path = path
+
+
+class DocUtilsParseError(Diagnostic):
+    severity = Diagnostic.Level.warning
+
+
+class ErrorParsingYAMLFile(Diagnostic):
+    severity = Diagnostic.Level.error
+
+    def __init__(
+        self,
+        path: Path,
+        reason: str,
+        start: Union[int, Tuple[int, int]],
+        end: Union[None, int, Tuple[int, int]] = None,
+    ) -> None:
+        super().__init__(f"Error parsing YAML file {str(path)}: {reason}", start, end)
+        self.path = path
+        self.reason = reason
+
+
+class InvalidLiteralInclude(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class SubstitutionRefError(Diagnostic):
+    severity = Diagnostic.Level.error
+
+
+class VariableNotDeclaredConstant(Diagnostic):
```

Comment:
[nit] Similarly, we could shorten this class by calling it `ConstantNotDeclared`, but either way 😄

Reply:
Ah, actually good call: that's a better name!